File: native/src/label.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

// Radix and enum metadata needed to format a multi-bit value into its pill
// label. Mirrors the renderer's Radix type + per-row enum maps.
// `@"enum"` formats via the per-row enum table (value→label), falling back to
// hex for unmatched values and X/Z bits. `dec` is unsigned, `sdec` two's-complement
// signed (both share the divmod path; sdec negates the magnitude first).
pub const Radix = enum { bin, hex, dec, sdec, @"enum" };
pub const EnumEntry = struct { value: u32, label: []const u8 };

const HEX_UPPER = "0123456789ABCDEF";

// Bit `b` (0 = LSB) of a tide little-endian byte plane.
inline fn bitOf(bytes: []const u8, b: u32) u1 {
    const idx = b >> 3;
    if (idx >= bytes.len) return 0;
    return @intCast((bytes[idx] >> @intCast(b & 7)) & 1);
}

// Per-bit 2-state classification: '0' | '1' | 'X' | 'Z'. (m,l) = (x1,x0) bit.
inline fn bitChar(x0: []const u8, x1: []const u8, b: u32) u8 {
    const l = bitOf(x0, b);
    const m = bitOf(x1, b);
    if (m == 0) return if (l == 0) '0' else '1';
    return if (l == 0) 'X' else 'Z';
}

// 4 bits (LSB-first) of a tide byte plane starting at bit `lo`, read across the
// byte boundary. Bits past the plane read as 0 (zero-padded above the width).
inline fn nibble(plane: []const u8, lo: usize) u8 {
    const idx = lo >> 3;
    const sh: u4 = @intCast(lo & 7);
    var v: u16 = if (idx < plane.len) plane[idx] else 0;
    if (idx + 1 < plane.len) v |= @as(u16, plane[idx + 1]) << 8;
    return @intCast((v >> sh) & 0xF);
}

// Low 32 bits of the value (enum key), little-endian from the x0 byte plane.
fn lowWord(x0: []const u8) u32 {
    var w: u32 = 0;
    var b: u32 = 0;
    while (b < 4) : (b += 1) {
        if (b < x0.len) w |= @as(u32, x0[b]) << @intCast(b * 8);
    }
    return w;
}

// Decimal of an arbitrary-width little-endian value (x0 byte plane), via
// repeated divmod-by-10 over a mutable u32-word copy. Matches the JS BigInt
// path so widths > 32 print exactly. Only called on the all-defined dec path.
// `signed`: treat the value as two's complement — if the sign bit (MSB) is set,
// emit '-' and divmod the negated magnitude.
fn appendDecimal(out: *std.ArrayList(u8), gpa: Allocator, x0: []const u8, width: u32, signed: bool) !void {
    const words: u32 = (width + 31) / 32;
    var buf = try gpa.alloc(u32, words);
    defer gpa.free(buf);
    var w: u32 = 0;
    while (w < words) : (w += 1) {
        var word: u32 = 0;
        var b: u32 = 0;
        while (b < 4) : (b += 1) {
            const idx = w * 4 + b;
            if (idx < x0.len) word |= @as(u32, x0[idx]) << @intCast(b * 8);
        }
        buf[w] = word;
    }
    // Clear bits above `width` in the top word so masking/negation are exact.
    const top_bits: u5 = @intCast(width & 31);
    if (top_bits != 0) buf[words - 1] &= (@as(u32, 1) << top_bits) - 1;

    // Signed & negative: emit '-' and replace buf with its two's-complement
    // magnitude (~buf + 1, masked back to `width`) before the divmod below.
    if (signed and bitOf(x0, width - 1) == 1) {
        try out.append(gpa, '-');
        var carry: u64 = 1;
        var i: u32 = 0;
        while (i < words) : (i += 1) {
            const s = @as(u64, ~buf[i]) + carry;
            buf[i] = @truncate(s);
            carry = s >> 32;
        }
        if (top_bits != 0) buf[words - 1] &= (@as(u32, 1) << top_bits) - 1;
    }

    // Collect digits least-significant first, then reverse.
    var digits: std.ArrayList(u8) = .empty;
    defer digits.deinit(gpa);
    while (true) {
        var rem: u64 = 0;
        var nonzero = false;
        var i: usize = words;
        while (i > 0) {
            i -= 1;
            const cur = (rem << 32) | buf[i];
            buf[i] = @intCast(cur / 10);
            rem = cur % 10;
            if (buf[i] != 0) nonzero = true;
        }
        try digits.append(gpa, @as(u8, '0') + @as(u8, @intCast(rem)));
        if (!nonzero) break;
    }
    var k: usize = digits.items.len;
    while (k > 0) {
        k -= 1;
        try out.append(gpa, digits.items[k]);
    }
}

inline fn bitmask(in: u8) u8 {
    const signed: i8 = @bitCast(in);
    return @bitCast(-signed);
}

// Format a multi-bit value into `out`. x0/x1 are tide's per-sample
// little-endian byte planes (value bits / unknown bits).
pub fn formatValue(
    out: *std.ArrayList(u8),
    gpa: Allocator,
    x0s: []const u8,
    x1s: []const u8,
    width: u32,
    radix: Radix,
    enums: []const EnumEntry,
) !void {
    switch (radix) {
        .bin => {
            try out.ensureUnusedCapacity(gpa, 2 + width);
            out.appendSliceAssumeCapacity("0b");

            var i: i64 = width - 1;
            while (i >= 0) : (i -= 1) {
                const byte = @as(usize, @intCast(i)) >> 3;
                const bit = @as(usize, @intCast(i)) & 0x7;
                const x0 = (x0s[byte] >> @intCast(bit)) & 0x1;
                const x1 = (x1s[byte] >> @intCast(bit)) & 0x1;

                // Start by choosing either '0' or 'X' depending on x1.
                var char: u8 = '0';
                char += ('X' - '0') & bitmask(x1);

                // Use x0 to shift char to '0' -> '1', or 'X' -> 'Y'.
                char += @as(u8, 1) & bitmask(x0);

                // If x0 & x1, shift 'Y' -> 'Z'.
                char += @as(u8, 1) & bitmask(x0 & x1);

                out.appendAssumeCapacity(char);
            }

            return;
        },
        .hex => {
            // Nibbles MSB-first, always 0x-prefixed, fixed width (no trim).
            const nibbles = (width + 3) / 4;
            try out.ensureUnusedCapacity(gpa, 2 + nibbles);
            out.appendSliceAssumeCapacity("0x");

            var hi: i64 = @as(i64, @intCast(width)) - 1;
            while (hi >= 0) : (hi -= 4) {
                const lo: usize = @intCast(@max(hi - 3, 0));
                // Mask off bits above the nibble (the bottom nibble may be < 4
                // bits wide when the width isn't a multiple of 4).
                const nbits: u3 = @intCast(hi - @as(i64, @intCast(lo)) + 1);
                const mask: u8 = (@as(u8, 1) << nbits) - 1;
                const x0n = nibble(x0s, lo) & mask;
                const x1n = nibble(x1s, lo) & mask;

                // Any unknown bit (x1 set) makes the whole nibble X or Z; x0 of
                // the unknown bits picks Z (set) over X. Otherwise x0 is the hex
                // digit. Branchless: build both, select with an all-ones mask.
                const unknown = bitmask(@intFromBool(x1n != 0));
                const isz = bitmask(@intFromBool((x0n & x1n) != 0));
                const hex_char = HEX_UPPER[x0n];
                const unk_char = @as(u8, 'X') + (@as(u8, 'Z' - 'X') & isz);
                out.appendAssumeCapacity((hex_char & ~unknown) | (unk_char & unknown));
            }
            return;
        },
        else => {}, // dec, enum
    }

    // dec / enum: whole-value x/z presence, OR-reduced per byte (each byte holds
    // distinct bits, so (m & ~l)/(m & l) never cross-contaminate).
    var has_x = false;
    var has_z = false;
    {
        var i: usize = 0;
        while (i < x1s.len) : (i += 1) {
            const m = x1s[i];
            const l: u8 = if (i < x0s.len) x0s[i] else 0;
            if ((m & ~l) != 0) has_x = true;
            if ((m & l) != 0) has_z = true;
        }
    }

    // Enum: match the table on fully-defined values; otherwise format as hex.
    if (radix == .@"enum") {
        if (!has_x and !has_z) {
            const key = lowWord(x0s);
            for (enums) |e| {
                if (e.value == key) {
                    try out.appendSlice(gpa, e.label);
                    return;
                }
            }
        }
        return formatValue(out, gpa, x0s, x1s, width, .hex, &.{});
    }

    // Decimal.
    if (has_x or has_z) {
        if (width == 1) {
            try out.append(gpa, bitChar(x0s, x1s, 0));
            return;
        }
        // Classify the whole value (any defined bit / any X / any Z).
        var any_x = false;
        var any_z = false;
        var any_def = false;
        {
            var b: u32 = 0;
            while (b < width) : (b += 1) {
                switch (bitChar(x0s, x1s, b)) {
                    'X' => any_x = true,
                    'Z' => any_z = true,
                    else => any_def = true,
                }
            }
        }
        // Uniformly-unknown reads better as a bare "X"/"Z" than a digit string.
        if (!any_def and !(any_x and any_z)) {
            try out.append(gpa, if (any_z) 'Z' else 'X');
            return;
        }
        // Mixed value: per-bit binary, MSB-first.
        try out.appendSlice(gpa, "0b");
        var b: i64 = @as(i64, width) - 1;
        while (b >= 0) : (b -= 1) {
            try out.append(gpa, bitChar(x0s, x1s, @intCast(b)));
        }
        return;
    }

    if (width == 1) {
        // 1-bit two's complement: bit set is -1 (signed) or 1 (unsigned).
        if (radix == .sdec and bitOf(x0s, 0) == 1) {
            try out.appendSlice(gpa, "-1");
        } else {
            try out.append(gpa, '0' + @as(u8, bitOf(x0s, 0)));
        }
        return;
    }
    try appendDecimal(out, gpa, x0s, width, radix == .sdec);
}

File: native/src/label.rs
//! Value labels for waveform rows.
//!
//! Formats one sample of tide's byte planes into the ASCII a pill or a boolean
//! line shows. A sample is bounds plus a high impedance mask, little-endian:
//! `0` is `(min 0, max 0)`, `1` is `(1, 1)`, `x` is `(0, 1)` and `z` is `(0, 1)`
//! with its `z` bit set. A bit is unknown exactly where the bounds disagree, so
//! `min ^ max` is the unknown mask and `min` is the value with every unknown bit
//! read as zero.
//!
//! Output is appended to a caller-owned buffer and must stay byte-for-byte
//! identical to the renderer's `wave/value.ts` `formatSegmentValue`, which
//! formats the same values CPU-side for the active-signal value column.
use tide_core::metadata::Width;

/// How to render a value. `Enum` looks the value up in the row's table and falls
/// back to hex for an unmatched value or any unknown bit. `Dec` is unsigned,
/// `Sdec` two's complement. `Boolean` renders on the single (line) pipeline but
/// still carries a `true`/`false`/`x` label.
#[derive(Clone, Copy, PartialEq, Eq, Debug, Default)]
pub enum Radix {
    #[default]
    Bin,
    Hex,
    Dec,
    Sdec,
    Enum,
    Boolean,
}

impl Radix {
    /// Parses the renderer's spelling. Anything unrecognized is binary, matching
    /// the scalar default in `scene.ts`'s pack specs.
    pub fn parse(name: &str) -> Radix {
        match name {
            "hex" => Radix::Hex,
            "dec" => Radix::Dec,
            "sdec" => Radix::Sdec,
            "enum" => Radix::Enum,
            "boolean" => Radix::Boolean,
            _ => Radix::Bin,
        }
    }

    /// Whether this radix renders as a pill (multi pipeline) rather than a
    /// high/low line. Routing is format-driven, not width-driven.
    pub fn is_multi(self) -> bool {
        !matches!(self, Radix::Bin | Radix::Boolean)
    }

    /// Whether a segment of this radix carries a value label at all. Binary
    /// (plain, clock and reset rows) is the only unlabeled format.
    pub fn is_labeled(self) -> bool {
        self != Radix::Bin
    }
}

/// One row's integer to name mapping, as the sidecar records it.
pub struct EnumEntry {
    pub value: u32,
    pub label: String,
}

const HEX_UPPER: &[u8; 16] = b"0123456789ABCDEF";

/// Bit `b` (0 = LSB) of a little-endian byte plane. Bits past the plane read as
/// zero, which is what the planes are padded with above the declared width.
fn bit_of(bytes: &[u8], b: u32) -> u8 {
    let index = (b >> 3) as usize;
    if index >= bytes.len() {
        return 0;
    }
    (bytes[index] >> (b & 7)) & 1
}

/// Per-bit four-state classification. The bounds disagree exactly on an unknown
/// bit, and the `z` plane then says which unknown it is.
fn bit_char(min: &[u8], max: &[u8], z: &[u8], b: u32) -> u8 {
    let (low, high) = (bit_of(min, b), bit_of(max, b));
    let unknown = low ^ high;
    let pick = if unknown == 1 { bit_of(z, b) } else { low };
    STATE_CHARS[(unknown << 1 | pick) as usize]
}

/// The four states, indexed by `unknown << 1` plus the value bit when known and
/// the high impedance bit when not.
const STATE_CHARS: [u8; 4] = [b'0', b'1', b'X', b'Z'];

/// Four bits of a plane starting at bit `lo`, read across the byte boundary.
fn nibble(plane: &[u8], lo: usize) -> u8 {
    let index = lo >> 3;
    let shift = lo & 7;
    let mut value = u16::from(plane.get(index).copied().unwrap_or(0));
    value |= u16::from(plane.get(index + 1).copied().unwrap_or(0)) << 8;
    ((value >> shift) & 0xf) as u8
}

/// Low 32 bits of the value, which is the enum table's key.
fn low_word(min: &[u8]) -> u32 {
    let mut word = 0;
    for (index, &byte) in min.iter().take(4).enumerate() {
        word |= u32::from(byte) << (index * 8);
    }
    word
}

/// `1 -> 0xff`, `0 -> 0x00`, for branchless selection.
fn bitmask(input: u8) -> u8 {
    (input as i8).wrapping_neg() as u8
}

/// Decimal of an arbitrary width little-endian value, by repeated divmod by ten
/// over a mutable word copy. Matches the renderer's `BigInt` path, so widths past
/// 32 bits print exactly. Only reached when every bit is known.
///
/// `signed` reads the value as two's complement: a set sign bit emits `-` and the
/// magnitude is negated before the divmod.
fn append_decimal(out: &mut Vec<u8>, min: &[u8], width: Width, signed: bool) {
    let words = width.div_ceil(32) as usize;
    let mut buf = vec![0u32; words];
    for (index, word) in buf.iter_mut().enumerate() {
        for byte in 0..4 {
            if let Some(&value) = min.get(index * 4 + byte) {
                *word |= u32::from(value) << (byte * 8);
            }
        }
    }

    // Clear bits above `width` in the top word so masking and negation are exact.
    let top_bits = width & 31;
    let mask_top = |buf: &mut [u32]| {
        if top_bits != 0 {
            buf[words - 1] &= (1u32 << top_bits) - 1;
        }
    };
    mask_top(&mut buf);

    if signed && bit_of(min, width - 1) == 1 {
        out.push(b'-');
        let mut carry = 1u64;
        for word in &mut buf {
            let sum = u64::from(!*word) + carry;
            *word = sum as u32;
            carry = sum >> 32;
        }
        mask_top(&mut buf);
    }

    // Least significant digit first, then reversed in place.
    let start = out.len();
    loop {
        let mut remainder = 0u64;
        let mut nonzero = false;
        for word in buf.iter_mut().rev() {
            let current = (remainder << 32) | u64::from(*word);
            *word = (current / 10) as u32;
            remainder = current % 10;
            nonzero |= *word != 0;
        }
        out.push(b'0' + remainder as u8);
        if !nonzero {
            break;
        }
    }
    out[start..].reverse();
}

/// Appends the label for one sample. `min`, `max` and `z` are that sample's
/// per-plane byte runs, exactly `ceil(width / 8)` bytes each.
pub fn format_value(
    out: &mut Vec<u8>,
    min: &[u8],
    max: &[u8],
    z: &[u8],
    width: Width,
    radix: Radix,
    enums: &[EnumEntry],
) {
    match radix {
        Radix::Bin => {
            out.reserve(2 + width as usize);
            out.extend_from_slice(b"0b");
            for bit in (0..width).rev() {
                out.push(bit_char(min, max, z, bit));
            }
            return;
        }
        Radix::Hex => {
            // Nibbles are aligned to the LSB, so the *top* nibble carries the
            // leftover `width % 4` bits: a 7-bit 0x7b prints "7B", not "F3".
            // Always prefixed, never trimmed. Keep in lockstep with value.ts.
            let nibbles = width.div_ceil(4) as usize;
            out.reserve(2 + nibbles);
            out.extend_from_slice(b"0x");

            let mut lo = ((width - 1) / 4 * 4) as usize;
            loop {
                let hi_bit = (lo + 3).min(width as usize - 1);
                let bits = hi_bit - lo + 1;
                let mask = ((1u16 << bits) - 1) as u8;
                let low = nibble(min, lo) & mask;
                let unknown_bits = (low ^ nibble(max, lo)) & mask;

                // Any unknown bit poisons the whole nibble, and one of those bits
                // being high impedance makes it Z rather than X.
                let unknown = bitmask(u8::from(unknown_bits != 0));
                let is_z = bitmask(u8::from((unknown_bits & nibble(z, lo)) != 0));
                let hex = HEX_UPPER[low as usize];
                let unknown_char = b'X' + ((b'Z' - b'X') & is_z);
                out.push((hex & !unknown) | (unknown_char & unknown));

                if lo < 4 {
                    break;
                }
                lo -= 4;
            }
            return;
        }
        Radix::Boolean => {
            // Whole-plane reduction: the planes are zero above the declared width.
            if min.iter().zip(max).any(|(&low, &high)| low != high) {
                out.extend_from_slice(b"x");
            } else if min.iter().any(|&byte| byte != 0) {
                out.extend_from_slice(b"true");
            } else {
                out.extend_from_slice(b"false");
            }
            return;
        }
        Radix::Dec | Radix::Sdec | Radix::Enum => {}
    }

    // Whole-value x/z presence, reduced per byte. Each byte holds distinct bits,
    // so the two tests never cross-contaminate.
    let mut has_x = false;
    let mut has_z = false;
    for (index, (&low, &high)) in min.iter().zip(max).enumerate() {
        let unknown = low ^ high;
        let high_impedance = z.get(index).copied().unwrap_or(0);
        has_x |= (unknown & !high_impedance) != 0;
        has_z |= (unknown & high_impedance) != 0;
    }

    if radix == Radix::Enum {
        if !has_x && !has_z {
            let key = low_word(min);
            if let Some(entry) = enums.iter().find(|entry| entry.value == key) {
                out.extend_from_slice(entry.label.as_bytes());
                return;
            }
        }
        return format_value(out, min, max, z, width, Radix::Hex, &[]);
    }

    if has_x || has_z {
        if width == 1 {
            out.push(bit_char(min, max, z, 0));
            return;
        }

        let (mut any_x, mut any_z, mut any_defined) = (false, false, false);
        for bit in 0..width {
            match bit_char(min, max, z, bit) {
                b'X' => any_x = true,
                b'Z' => any_z = true,
                _ => any_defined = true,
            }
        }

        // A uniformly unknown value reads better as a bare X or Z than as a
        // digit string.
        if !(any_defined || any_x && any_z) {
            out.push(if any_z { b'Z' } else { b'X' });
            return;
        }

        // Mixed: per-bit binary, most significant first.
        out.extend_from_slice(b"0b");
        for bit in (0..width).rev() {
            out.push(bit_char(min, max, z, bit));
        }
        return;
    }

    if width == 1 {
        // One-bit two's complement: a set bit is -1 signed, 1 unsigned.
        if radix == Radix::Sdec && bit_of(min, 0) == 1 {
            out.extend_from_slice(b"-1");
        } else {
            out.push(b'0' + bit_of(min, 0));
        }
        return;
    }

    append_decimal(out, min, width, radix == Radix::Sdec);
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Planes for a value written most significant bit first, in the four-state
    /// spelling the oracle uses. Building tests from text rather than hand-packed
    /// planes keeps them readable and survives a change of encoding.
    fn planes(bits: &str) -> (Vec<u8>, Vec<u8>, Vec<u8>) {
        let bytes = bits.len().div_ceil(8);
        let (mut min, mut max, mut z) = (vec![0u8; bytes], vec![0u8; bytes], vec![0u8; bytes]);
        for (index, digit) in bits.chars().rev().enumerate() {
            let (byte, shift) = (index / 8, index % 8);
            let (low, high, impedance) = match digit {
                '0' => (0, 0, 0),
                '1' => (1, 1, 0),
                'x' => (0, 1, 0),
                'z' => (0, 1, 1),
                other => panic!("not a four-state digit: {other}"),
            };
            min[byte] |= low << shift;
            max[byte] |= high << shift;
            z[byte] |= impedance << shift;
        }
        (min, max, z)
    }

    /// Formats a value written as four-state text; the width is its length.
    fn format(bits: &str, radix: Radix) -> String {
        let (min, max, z) = planes(bits);
        let mut out = Vec::new();
        format_value(&mut out, &min, &max, &z, bits.len() as Width, radix, &[]);
        String::from_utf8(out).unwrap()
    }

    #[test]
    fn binary_prints_every_declared_bit() {
        assert_eq!("0b0000", format("0000", Radix::Bin));
        assert_eq!("0b1010", format("1010", Radix::Bin));
        assert_eq!("0bXZXZ", format("xzxz", Radix::Bin));
        // Wider than one byte, most significant bit first.
        assert_eq!("0b0000000100000000", format("0000000100000000", Radix::Bin));
    }

    #[test]
    fn hex_nibbles_align_to_the_lsb() {
        // A 7-bit 0x7b keeps its low nibble whole: the leftover 3 bits are the
        // top nibble, so this is "7B" and never "F3".
        assert_eq!("0x7B", format("1111011", Radix::Hex));
        assert_eq!("0x0A", format("00001010", Radix::Hex));
        assert_eq!("0x1AB", format("110101011", Radix::Hex));
    }

    #[test]
    fn one_unknown_bit_poisons_its_whole_nibble() {
        // The high nibble is untouched either way.
        assert_eq!("0x1X", format("0001001x", Radix::Hex));
        assert_eq!("0x1Z", format("0001001z", Radix::Hex));
        // A nibble holding both reads as Z: high impedance is the louder fault.
        assert_eq!("0x1Z", format("000100xz", Radix::Hex));
    }

    #[test]
    fn decimal_spans_widths_past_a_word() {
        assert_eq!("0", format("00000000", Radix::Dec));
        assert_eq!("255", format("11111111", Radix::Dec));
        // 2^64, which no u64 path could print.
        let wide = format!("1{}", "0".repeat(64));
        assert_eq!("18446744073709551616", format(&wide, Radix::Dec));
    }

    #[test]
    fn signed_decimal_negates_the_magnitude() {
        assert_eq!("255", format("11111111", Radix::Dec));
        assert_eq!("-1", format("11111111", Radix::Sdec));
        assert_eq!("-128", format("10000000", Radix::Sdec));
        assert_eq!("127", format("01111111", Radix::Sdec));
        // One bit wide is its own case: the sign bit is the only bit.
        assert_eq!("-1", format("1", Radix::Sdec));
        assert_eq!("1", format("1", Radix::Dec));
    }

    #[test]
    fn decimal_falls_back_to_bits_only_when_the_value_is_mixed() {
        // Uniformly unknown reads as a bare letter.
        assert_eq!("X", format("xxxxxxxx", Radix::Dec));
        assert_eq!("Z", format("zzzzzzzz", Radix::Dec));
        // One unknown bit among known ones falls back to per-bit binary.
        assert_eq!("0b0000000Z", format("0000000z", Radix::Dec));
        // Both an X and a Z present is mixed even with no defined bit.
        assert_eq!("0bZX", format("zx", Radix::Dec));
    }

    #[test]
    fn booleans_reduce_the_whole_value() {
        assert_eq!("false", format(&"0".repeat(16), Radix::Boolean));
        assert_eq!("true", format("1000000000000000", Radix::Boolean));
        assert_eq!("x", format("000000z000000000", Radix::Boolean));
    }

    #[test]
    fn enums_match_on_the_low_word_and_fall_back_to_hex() {
        let table = [
            EnumEntry {
                value: 0,
                label: "IDLE".into(),
            },
            EnumEntry {
                value: 2,
                label: "RUN".into(),
            },
        ];
        let format = |bits: &str| {
            let (min, max, z) = planes(bits);
            let mut out = Vec::new();
            format_value(&mut out, &min, &max, &z, 4, Radix::Enum, &table);
            String::from_utf8(out).unwrap()
        };

        assert_eq!("IDLE", format("0000"));
        assert_eq!("RUN", format("0010"));
        // No table entry, and any unknown bit, both fall through to hex.
        assert_eq!("0x3", format("0011"));
        assert_eq!("0xZ", format("zzzz"));
    }
}
